File: data-structures-algorithms/leetcode/go/42.go
func trap(height []int) int {
	left := 0
	right := len(height) - 1
	// 左边最大高度
	leftMax := 0
	// 右边最大高度
	rightMax := 0
	totalWater := 0

	for left < right {
		if height[right] > height[left] {
			if height[left] < leftMax {
				// 该列能够存储的最多水
				totalWater += leftMax - height[left]
			} else {
				leftMax = height[left]
			}
			left++
		} else {
			if height[right] < rightMax {
				// 该列能够存储的最多水
				totalWater += rightMax - height[right]
			} else {
				rightMax = height[right]
			}
			right--
		}
	}

	return totalWater
}

File: data-structures-algorithms/leetcode/js/42.js
function trap(height) {
  const stack = []
  let totalWater = 0

  for (let i = 0; i < height.length; i++) {
    while (stack.length && height[i] > height[stack[stack.length - 1]]) {
      const top = stack.pop()
			// 如果左边没有了则收集不到雨水
      if (!stack.length) {
        continue
      }
      const left = stack[stack.length - 1]
      const currentWidth = i - left - 1
      const currentHeight = Math.min(height[left], height[i]) - height[top]
      totalWater += currentHeight * currentWidth
    }

    stack.push(i)
  }

  return totalWater
}
